src/Chat.tsx

```diff
@@ -74,6 +78,8 @@ export class Chat extends React.Component<ChatProps, {}> {
 
         konsole.log("BotChat.Chat props", props);
 
+        this.handleKeyDownCapture = this.handleKeyDownCapture.bind(this);
```

Comment:
is this necessary?

Reply:
Yes, look at this [tutorial](https://facebook.github.io/react/docs/handling-events.html) from React. For ES6 class, we should call `.bind()` to make sure `this` points to the component itself.

Some people do it in `render()`, but it hurts performance.

Reply:
oh yes of course. Can you use a lambda instead? We already have the lambda style in https://github.com/Microsoft/BotFramework-WebChat/blob/master/src/MessagePane.tsx#L59

Reply:
I could use that, but it's still a perf hit. Because every time it `render()`, we will create a new lambda function, and default `shouldComponentUpdate()` won't be able to figure out that both lambda functions actually land on the same event handler. Thus, a wasted render will be detected thru `React.addons.Perf.printWasted()` call.

Reply:
oh, well then we should probably convert the existing lambdas... but not in this PR.

Reply:
Reassigning the instance variable using `bind` is both clever and ugly in the extreme. I much prefer creating a second instance variable:

```ts
_handleKeyDownCapture = (evt: React.KeyboardEvent<HTMLDivElement>) => this.handleKeyDownCapture(evt);

...

onKeyDownCapture={ this._handleKeyDownCapture }
```

Prettier, with the same perf. (We do this pattern elsewhere, or at least we did.)

More importantly, @compulim, I think you have an important misconception here. There is no such thing as "default `shouldComponentUpdate()`". By default there is no `shouldComponentUpdate` - you have to add it yourself. It is true that a simple and common implementation is to just compare the value of `props`, in which case you're right, you want to avoid creating a new lambda on every call.

Reply:
Agree that we should just go with an arrow function. Less code and less confusing. 👍

Reply:
@billba I was using `PureComponent` for my other projects so I didn't notice we wasn't using it.

Will change to arrow second variable instead of `.bind()` to existing.

Reply:
You're right - if you're using PureComponent then you *do* get a default `shouldComponentUpdate` that just looks at props - and we do use this for `Carousel` and `CarouselAttachment`.